### PR TITLE
[CenterOfMass]: Add bounding box

### PR DESCRIPTION
**Pull request by chupins**
Draw a bounding box around all selected object when you choose to look at the center of masses

### CODE REVIEW

**Comment by s-quirin** [comment]
@chupins We better should test it internally first and not confuse the user with numerous updates. Please check out [my fork](https://github.com/s-quirin/FreeCAD-macros) with Bugfixes.

**Comment by chupins** [comment]
@s-quirin, I merged your changes in [my fork](https://github.com/chupins/FreeCAD-macros).
And then, it's in the present merge request

File: Information/CenterOfMass.FCMacro [hunk 1]
```diff
@@ -250,11 +270,11 @@ class CenterofmassWidget(QtGui.QWidget):
         label_comboUnit.setAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
 
         comboUnitDensity = QtGui.QComboBox(toolTip='Unit of density')
-        comboUnitDensity.addItems(['kg/m³',
-                                   'g/dm³', 'g/cm³',
-                                   'mg/mm³',
-                                   'oz/in³',
-                                   'lb/in³', 'lb/ft³', 'lb/yd³'])
+        comboUnitDensity.addItems(['kg/m^3',
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:273]: The original was nicer. Why this change?

File: Information/CenterOfMass.FCMacro [hunk 2]
```diff
@@ -363,30 +377,51 @@ class CenterofmassWidget(QtGui.QWidget):
...
         showCoM = QtGui.QCheckBox(toolTip='Show center of mass')
         if self.doc.getObject('CenterOfMass'):
             showCoM.setChecked(True)
         showCoM.setIcon(QtGui.QIcon(':/icons/Std_ToggleVisibility.svg'))
         showCoM.setIconSize(g_icon_size)
         showCoM.stateChanged.connect(self.on_stateChanged_showCoM)
 
-        self.changeRadius = QtGui.QSlider(QtGui.Qt.Vertical)
+        self.switchVolumMass = QtGui.QPushButton('mass')
+        self.switchVolumMass.setToolTip('switch radius of spheres to represent relative masses or volumes')
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:397]: Wouldn't `Toggle` be better? In any case, the sentence should start with a capital letter.

File: Information/CenterOfMass.FCMacro [hunk 3]
```diff
@@ -589,8 +622,8 @@ class CenterofmassWidget(QtGui.QWidget):
     def store_prefered_units(self, txt):
         """Get units set as preference"""
         self.unitForD_text = txt
-        self.unitForD = self.unitForD_text.replace('³', '^3')         # density
-        self.unitForL = self.unitForD_text.split('/')[1].rstrip('³')  # length
+        self.unitForD = self.unitForD_text.replace('Â³', '^3')         # density
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:625]: I see a strange sign in Github (`Â`). Is it normal?

File: Information/CenterOfMass.FCMacro [hunk 4]
```diff
@@ -761,10 +796,22 @@ class CenterofmassWidget(QtGui.QWidget):
             plane.ViewObject.Transparency = 100
             CoMObjs.addObject(plane)
 
+        ### BoundingBox
+        BBoxSolid = self.doc.addObject("Part::Box","CoMBBox")
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:800]: Please use single-quotes.

File: Information/CenterOfMass.FCMacro [hunk 5]
```diff
@@ -761,10 +796,22 @@ class CenterofmassWidget(QtGui.QWidget):
...
+        BBoxSolid.Placement.Base = self.boundBox.Center.sub(app.Vector(boundBoxL).multiply(0.5))
+        BBoxSolid.Length = boundBoxL[0]
+        BBoxSolid.Width  = boundBoxL[1]
+        BBoxSolid.Height = boundBoxL[2]
+        BBoxSolid.ViewObject.Transparency = 90
+        BBoxSolid.ViewObject.ShapeColor = (0.5, 0.5, 0.5)
+#        BBoxSolid.ViewObject.BoundingBox = True
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:807]: Please activate or remove this line.

File: Information/CenterOfMass.FCMacro [hunk 6]
```diff
@@ -824,16 +889,31 @@ class CenterofmassWidget(QtGui.QWidget):
...
-            color = QtGui.QColor.fromHsvF(h, 1, 1)
+                drel = math.acos(1-2*drel) / math.pi    # non-linear relation to stretch the middle color
+                index = int((COLORMAP_BINS-1)*drel) # position of the density on the colormap
+            color = QtGui.QColor.fromRgbF(colormap[index][0], colormap[index][1], colormap[index][2])
             self.doc.getObject(g_sel[sol].Name).ViewObject.ShapeColor = color.getRgbF()
             self.solids[sol].spinDens.setStyleSheet('QDoubleSpinBox {background-color: %s}' %color.name())
+            if self.doc.getObject('CenterOfMass'):
+                self.doc.getObject('CoM_' + g_sel[sol].Name).ViewObject.ShapeColor = color.getRgbF()
+
+    def make_gradient_from_three_colors(self, color_begin, color_middle, color_end) :
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:904]: This function doesn't use `self`, so it should be written as a free function.

**Review by galou — CHANGES_REQUESTED** [review]
Thanks for the changes! I didn't test them but just had a look at the code. There are a few things that can be done to improve it, please consider them. Thanks!

File: Information/CenterOfMass.FCMacro (hunk 1 above)
Comment (chupins) [reply Information/CenterOfMass.FCMacro:273]: I have a strange (Â) that is appearing sometimes. The ^3 seems more "stable"

**Comment by chupins** [comment]
Sorry but I can't see anything above that clearly explains what's broken !
I don't know if it's the proper way to do it, but your MR contains a lot of thing that are already pending here.

**Comment by chupins** [comment]
A lot of changes made by s-quirin in #129 were already made here.
So I manually merged its version in here and took @galou remarks into account.

@s-quirin can you check if you find any problem ?
When I do a diff between your version and this one, I can't find any fundamental difference (the only things are my new colormaps and the mass/volume button)

File: Information/CenterOfMass.FCMacro (hunk 3 above)
Comment (chupins) [reply Information/CenterOfMass.FCMacro:638]: It look like passing through different os the ³ changes in Â³

Comment (galou) [reply Information/CenterOfMass.FCMacro:638]: I'm sorry but I don't believe this. This is not a problem with the OS but with your editor. You should set the endocing to UTF-8 without BOM.

**Comment by galou** [comment]
@chupins: @s-quirin removed some files in his PR, can you do the same if these files are indeed not needed.

I'll wait for @s-quirin's approval before merging.

**Comment by galou** [comment]
@chupins: @s-quirin removed some files in his PR, can you do the same if these files are indeed not needed.

I'll wait for @s-quirin's approval before merging.

File: Information/CenterOfMass.FCMacro (hunk 3 above)
Comment (chupins) [reply Information/CenterOfMass.FCMacro:638]: You're right, it's the text editor ! (but some time I swtich between windows and linux and I also switch between text editors...)

**Comment by chupins** [comment]
You're right again, I forgot to delete these unecessary files.
We'll see with @s-quirin how we merge the PR and the #129.

**Comment by galou** [comment]
@chupins @s-quirin I don't know whether I should merge this PR, #129, or both. Can you please express yourself here? Thanks!

**Comment by s-quirin** [comment]
> @chupins @s-quirin I don't know whether I should merge this PR, #129, or both. Can you please express yourself here? Thanks!

We couldn't come to an agreement yet because of holidays / lack of time. #128 mainly contains two controversial changes that emerged in my own test and with end users:
- The originally worked out color palette was replaced by a pre-selection from matplotlib. I revived the built-in colormap and added a setting for the user to add any colormap to the pre-selection.
- A feature in which the visualization is not about the mass but about the volume. The user can confuse both. A clear UI solution is not easy and should be postponed to a next version.

I asked @chupins to fix them in this PR or to withdraw it and go with #129.

**Comment by chupins** [comment]
@s-quirin has added everything on #129 .
We prefer canceling here and ask if #129 to be accepted.

**Pull request closed without merging**